### PR TITLE
[Startup][Refactor][Backport] Memory allocation fail handler + init step 1 refactored.

### PR DESCRIPTION
**Pull request by furszy**
* OS memory allocation fail handler.
* OS signal handler registration method created to remove code duplication.
* AppInitBasicSetup() method created, better code organization of the wallet initialization setup step (step 1).
* Stop loading block indexes if wallet shutdown was requested.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1290,6 +1291,12 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
             if (pindex->nHeight % 100 == 0 && dProgressTip - dProgressStart > 0.0)
                 ShowProgress(_("Rescanning..."), std::max(1, std::min(99, (int)((Checkpoints::GuessVerificationProgress(pindex, false) - dProgressStart) / (dProgressTip - dProgressStart) * 100))));
 
+            if (fromStartup) {
+                if (ShutdownRequested()) {
+                    return -1;
+                }
+            }
+
```
Comment (random-zebra) [inline src/wallet/wallet.cpp]: styling nit: `if (fromStartup && ShutdownRequested())` ?

File: src/init.cpp [hunk 2]
```diff
@@ -1746,7 +1746,10 @@ bool AppInit2()
             uiInterface.InitMessage(_("Rescanning..."));
             LogPrintf("Rescanning last %i blocks (from block %i)...\n", chainActive.Height() - pindexRescan->nHeight, pindexRescan->nHeight);
             const int64_t nWalletRescanTime = GetTimeMillis();
-            pwalletMain->ScanForWalletTransactions(pindexRescan, true);
+            if (pwalletMain->ScanForWalletTransactions(pindexRescan, true, true) == -1) {
+                LogPrintf("Shutdown requested over the txs scan. Exiting.\n");
+                return false;
+            }
```
Comment (random-zebra) [inline src/init.cpp]: styling nit: `return error("Shutdown requested over the txs scan. Exiting.")` (no need for line terminator `\n`)

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (furszy) [reply src/wallet/wallet.cpp]: done

File: src/init.cpp (hunk 2 above)
Comment (furszy) [reply src/init.cpp]: done

**Pull request merged**